mir-ci/mir_ci/robot_libraries/Screencopy.py

```diff
@@ -0,0 +1,169 @@
+        """
+        await self.connect()
+        regions = []
+        last_frame_count = 0
+        matched = False
+        screenshot = None
+        end_time = time.time() + float(timeout)
+        while time.time() < end_time:
+            if self.frame_count == last_frame_count:
+                await asyncio.sleep(0)
+            else:
+                last_frame_count = self.frame_count
+                try:
+                    screenshot = self.grab_screenshot()
+                    regions = self._rpa_images.find_template_in_image(
+                        screenshot,
+                        template,
+                        tolerance=self.TOLERANCE,
+                    )
+                    matched = True
+                    break
+                except (RuntimeError, ValueError, ImageNotFoundError):
+                    continue
+
+        if not matched:
+            self._log_failed_match(screenshot, template)
+            raise ImageNotFoundError
+
+        return [
+            {
+                "left": region.left,
+                "top": region.top,
+                "right": region.right,
+                "bottom": region.bottom,
+            }
+            for region in regions
+        ]
+
+    def grab_screenshot(self) -> Image.Image:
+        """Grab the latest captured frame from the display."""
+        assert self.shm_data is not None, "No SHM data available"
+
+        self.shm_data.seek(0)
+        data = self.shm_data.read()
+        size = (self.buffer_width, self.buffer_height)
+        stride = self.buffer_stride
+        image = Image.frombytes("RGBA", size, data, "raw", "RGBA", stride, -1)
+        b, g, r, a, *_ = image.split()
+        image = Image.merge("RGBA", (r, g, b, a))
+
+        return image
+
+    async def connect(self):
+        """Connect to the display."""
+        if not self.shm_data:
+            await super().connect()
+            # Wait for the first frame to become ready
+            while self.frame_count == 0:
+                await asyncio.sleep(0)
+
+    async def disconnect(self):
+        """Disconnect from the display."""
+        if self.shm_data:
+            await super().disconnect()
+
+    @staticmethod
+    def _to_base64(image: Image.Image) -> str:
+        """Convert Pillow Image to b64"""
+        im_file = BytesIO()
+        image.save(im_file, format="PNG")
+        im_bytes = im_file.getvalue()
+        im_b64 = base64.b64encode(im_bytes)
+        return im_b64.decode()
+
+    def _log_failed_match(self, screenshot, template):
+        """Log a failure with template matching."""
+        if not screenshot:
+            return
+
+        template_img = Image.open(template)
+        template_string = 'Template was:<br><img src="data:image/png;base64,' f'{self._to_base64(template_img)}" /><br>'
+        image_string = 'Image was:<br><img src="data:image/png;base64,' f'{self._to_base64(screenshot)}" />'
```

Comment:
Just some HTML sugar:

```suggestion
        template_string = 'Template was:<br /><img src="data:image/png;base64,' f'{self._to_base64(template_img)}" /><br />'
        image_string = 'Image was:<br /><img src="data:image/png;base64,' f'{self._to_base64(screenshot)}" />'
```